src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/WamAdapters.cs

```diff
@@ -58,45 +60,44 @@ internal class WamAdapters
-                        wamPlugin.MapTokenRequestError(wamResponse.ResponseStatus, wamResponse.ResponseError?.ErrorCode ?? 0, isInteractive);
+                        $"Wam Plugin: {wamPlugin.GetType()}" +
+                        $" Error Code: {internalErrorCode}" +
+                        $" Error Message: {wamResponse.ResponseError?.ErrorMessage}" + 
+                        $" Internal Error Code: {internalErrorCode}" +
+                        $" Retry: {error.Item2}";
+                    throw new MsalUiRequiredException(errorCode, errorMessage, error.Item2);
+
+                case WebTokenRequestStatus.UserCancel: 
+                    throw new MsalClientException(MsalError.AuthenticationCanceledError, MsalErrorMessage.AuthenticationCanceled);
+
+                case WebTokenRequestStatus.ProviderError: 
+                    error = wamPlugin.MapTokenRequestError(wamResponse.ResponseStatus, wamResponse.ResponseError?.ErrorCode ?? 0, isInteractive);
+                    errorCode = error?.Item1;
+                    internalErrorCode = (wamResponse.ResponseError?.ErrorCode ?? 0).ToString(CultureInfo.InvariantCulture);
                     errorMessage =
                         $"{WamErrorPrefix} {wamPlugin.GetType()} \n" +
                         $" Error Code: {errorCode} \n" +
                         $" Error Message: {wamResponse.ResponseError?.ErrorMessage} \n"  +
+                        $" Internal Error Code: {internalErrorCode} \n" +
+                        $" Retry: {error?.Item2} \n" +
                         $" Possible causes: \n " +
                         $"- Invalid redirect uri - ensure you have configured the following url in the AAD portal App Registration: {GetExpectedRedirectUri(clientId)} \n" +
                         $"- No Internet connection \n" +
                         $"Please see https://aka.ms/msal-net-wam for details about Windows Broker integration";
 
-                internalErrorCode = (wamResponse.ResponseError?.ErrorCode ?? 0).ToString(CultureInfo.InvariantCulture);
-                    break;
+                    
+                    throw new MsalServiceException(errorCode, errorMessage, error == null ? false : error.Item2);
+                    
                 default:
-                    errorCode = MsalError.UnknownBrokerError;
                     internalErrorCode = wamResponse.ResponseError.ErrorCode.ToString(CultureInfo.InvariantCulture);
                     errorMessage = $"Unknown WebTokenRequestStatus {wamResponse.ResponseStatus} (internal error code {internalErrorCode})";
-                    break;
-            }
 
-            return new MsalTokenResponse()
-            {
-                Error = errorCode,
-                ErrorCodes = internalErrorCode != null ? new[] { internalErrorCode } : null,
-                ErrorDescription = errorMessage
-            };
+                    throw new MsalServiceException(MsalError.UnknownBrokerError, errorMessage);
```

Comment:
Since instead of returning MsalTokenResponse, we throw an exceptions, does this change the end-to-end experience? If not, do the same type of exceptions are thrown in flows and error codes are the same?

Reply:
Should not change the end to end experience as the exceptions are internally converted to MSAL response with error codes and error messages. The error code and error message does not change apart from the addition to IsRetryable.